### ISSUE TITLE
Best practices for developing an accessible theme

### ISSUE THREAD

**Issue by jonathanbossenger** [WordPress/Learn#2188]
# Details

- Topic description: Theme accessibility
- Audience (User, Developer, Designer, Contributor, etc.): Developer, Designer
- Learning objectives (What will the learner be able to do as a result of this content?): What is accessibility, why is it important, and how to ensure your theme is accessible.
- Content type (Online Workshop, Lesson, Course, Tutorial, or Lesson Plan): Tutorial
- WordPress version (optional): 
- Will you be creating this content? (Yes or No): Yes

### WordPress
- The [accessibility chapter](https://developer.wordpress.org/themes/functionality/accessibility/) of the Theme Handbook.
- The [Accessibility Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/) chapter of the Coding Standards Handbook.
- [Requirements](https://make.wordpress.org/themes/handbook/review/accessibility/) for `accessibility-ready` themes.
- WordPress Accessibility [GitHub repository](https://github.com/wpaccessibility).
- [Learn tutorial: How to use headings for accessibility and SEO](https://learn.wordpress.org/tutorial/using-headings-for-accessibility-and-seo/)
- [Theme accessibility Lesson Plan](https://learn.wordpress.org/lesson-plan/theme-accessibility/)

### Other
- TBD

#### Recommended plugins
  - [Accessibility Checker](https://wordpress.org/plugins/accessibility-checker)
  - [Editoria11y Accessibility Checker](https://wordpress.org/plugins/editoria11y-accessibility-checker)
  - [Sa11y](https://wordpress.org/plugins/sa11y)
  - [WP Tota11y](https://wordpress.org/plugins/wp-tota11y)

## Lesson creation checklist

- [x] Gather any relevant links to Support, Docs, or related material
- [x] Description and Objectives finalized
- [x] Lesson created and announced to the team for review
- [ ] Lesson reviewed
- [ ] Lesson video submitted and published to WPTV
- [ ] Lesson created on Learn.WordPress.org 
- [ ] Lesson video published to YouTube
- [ ] Lesson on Learn.WordPress.org updated with YouTube video
- [ ] Lesson published to Learn.WordPress.org

**Comment by ironnysh** [comment]
# Details


- Topic description: Theme accessibility
- Audience (User, Developer, Designer, Contributor, etc.): Developer, Designer
- Learning objectives (What will the learner be able to do as a result of this content?): What is accessibility, why is it important, and how to ensure your theme is accessible.
- Content type (Online Workshop, Lesson, Course, Tutorial, or Lesson Plan): Tutorial
- WordPress version (optional): 
- Will you be creating this content? (Yes or No): Yes

## Related Resources


### WordPress
- The [accessibility chapter](https://developer.wordpress.org/themes/functionality/accessibility/) of the Theme Handbook.
- The [Accessibility Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/) chapter of the Coding Standards Handbook.
- [Requirements](https://make.wordpress.org/themes/handbook/review/accessibility/) for `accessibility-ready` themes.
- WordPress Accessibility [GitHub repository](https://github.com/wpaccessibility).
- [Learn tutorial: How to use headings for accessibility and SEO](https://learn.wordpress.org/tutorial/using-headings-for-accessibility-and-seo/)
- [Theme accessibility Lesson Plan](https://learn.wordpress.org/lesson-plan/theme-accessibility/)

### Other
- TBD

#### Recommended plugins
  - [Accessibility Checker](https://wordpress.org/plugins/accessibility-checker)
  - [Editoria11y Accessibility Checker](https://wordpress.org/plugins/editoria11y-accessibility-checker)
  - [Sa11y](https://wordpress.org/plugins/sa11y)
  - [WP Tota11y](https://wordpress.org/plugins/wp-tota11y)

## Lesson creation checklist

- [ ] Gather any relevant links to Support, Docs, or related material
- [ ] Description and Objectives finalized
- [ ] Lesson created and announced to the team for review
- [ ] Lesson reviewed
- [ ] Lesson video submitted and published to WPTV
- [ ] Lesson created on Learn.WordPress.org 
- [ ] Lesson video published to YouTube
- [ ] Lesson on Learn.WordPress.org updated with YouTube video
- [ ] Lesson published to Learn.WordPress.org

**Comment by jonathanbossenger** [comment]
@ironnysh this is an excellent start 👏 

As I read through this, I am wondering, does it make sense for all this information to be in one lesson, or would it make sense to break it up into 2 - 3 separate lessons? My thinking is that the following two topics could be the focus of each lesson:

1. What is accessibility, and why is it important?
2. Developing a theme that is accessible.
3. Tools to test your theme for accessibility.

Let me know what you think of this idea.

**Comment by ironnysh** [comment]
Thanks, @jonathanbossenger!

You're right. But I'm worried most people would skip the first strictly theoretical video (what is accessibility, and why is it important?) and jump right to the tips and tricks without understanding the why.

What if we stick to the "traditional" division of _Developer_ vs. _User_ used in the Learning Pathways? This way, both audiences share the same intro, but the practical part of the tutorial changes: 

**User:**
1. What is accessibility, and why is it important? (50%)
2. Tools to test your _content_ for accessibility. (50%)

**Developer/designer:**
1. What is accessibility, and why is it important? (25%)
2. Best practices when developing/designing an accessible theme. (25%)
3. Tools to test your theme for accessibility. (50%)

What say you?

Here's a suggested outline:👇

## Intro
Block-based themes and FSE offer incredible flexibility, but with great power comes great responsibility :-)

Many content managers, developers, and designers who create or maintain Block-based themes aren't aware of accessibility aspects: semantic document outline, color contrast, form labels, links text, alt text for images, and more.
Some of these are required by law, others are considered best practices, but they all contribute to a better browsing experience for diverse audiences.

Some stats that highlight why it's important:
- 96.3% of websites scanned by [WebAIM](https://webaim.org/projects/million/) had WCAG failures.
- 96.1% of all errors detected fall into six categories (see above).
- TPGi lists the [most common accessibility errors in 2023](https://www.tpgi.com/the-top-accessibility-errors-found-in-2023/).
- WordPress is by far the most popular CMS, with [~43% market share](https://w3techs.com/technologies/details/cm-wordpress).

## Outline:
- Explain the importance of a11y (including its effects on SEO and, to some extent, performance).
- Mention the legal requirements as well as the ethical and commercial implications of non-compliance.
- Explore common problems.
- Show how to avoid/fix them—either when designing and developing the site or while creating content.
   - For the latter, we'll review a few a11y plugins, which can detect and help resolve issues:
      - [Accessibility Checker](https://wordpress.org/plugins/accessibility-checker)
      - [Editoria11y Accessibility Checker](https://wordpress.org/plugins/editoria11y-accessibility-checker)
      - [Sa11y](https://wordpress.org/plugins/sa11y)
      - [WP Tota11y](https://wordpress.org/plugins/wp-tota11y)

**Comment by jonathanbossenger** [comment]
> But I'm worried most people would skip the first strictly theoretical video (what is accessibility, and why is it important?) and jump right to the tips and tricks without understanding the why.

I know, and that can happen even if we add the content to a single video. At least if we do have the theory easily available, those who want to know the why will have it available to them.

The downside to having a longer video is the possibility of the viewer jumping ahead anyway.

There's no right or wrong answer though.

> What if we stick to the "traditional" division of Developer vs. User used in the Learning Pathways? This way, both audiences share the same intro, but the practical part of the tutorial changes:

@westnz as the person working on the User learning pathway, perhaps you'd like to share you thoughts on this question.

**Comment by jonathanbossenger** [comment]
@ironnysh I think the outline looks great, I look forward to seeing where you take this.

**Comment by westnz** [comment]
> _What if we stick to the "traditional" division of Developer vs. User used in the Learning Pathways? This way, both audiences share the same intro, but the practical part of the tutorial changes:_
> 
> @westnz as the person working on the User learning pathway, perhaps you'd like to share you thoughts on this question.

That would be very helpful. Would you then be able to pick up this lesson from the Intermediate User Pathway?: https://github.com/WordPress/Learn/issues/2201 @ironnysh. If you add a note to this issue, I can assign it to you.

**Comment by jonathanbossenger** [comment]
@westnz one thing I forgot to add, @ironnysh is interested in working on research and scripts, but might not want to work on actual video production, so this might be a collaborative lesson. However, maybe we can find someone else to record the video from the script.

**Comment by westnz** [comment]
Perfect. I am happy to collaborate or find someone else to do the recording. 👍 Thank you for contributing to the learning pathways @ironnysh.

**Comment by ironnysh** [comment]
Hi @jonathanbossenger, the [draft for the script](https://docs.google.com/document/d/1DqIqftBwFxMG36FAJ2nMTqMkO2YEVbc142Jfc0SmyLk/edit?usp=sharing) is ready for review. Looking forward to your feedback :-) 

-- cc @westnz

**Comment by jonathanbossenger** [comment]
@ironnysh excellent work. I think you've provided a great overview of A11y best practices.

I've left some comments in the doc for you.

**Comment by jonathanbossenger** [comment]
@ironnysh I think we're good to record this, would you agree?

**Comment by ironnysh** [comment]
-- @jonathanbossenger, yup! I added one more suggestion to the `focus` comment (forgot to include it before)

**Comment by jonathanbossenger** [comment]
I'll probably be in recording mode next week, so I'll tackle this then. I'll shout if I have any questions.

**Comment by jonathanbossenger** [comment]
Hi @ironnysh 

I'm prepared to record this lesson, and I have a question. 

In the section on headings, you describe a **Document Overview > Outline** option. 

Where is this located, as I'm not able to find it?

**Comment by ironnysh** [comment]
Hey @jonathanbossenger, here's a screenshot from the post editor: 

<img width="1024" alt="image" src="https://github.com/WordPress/Learn/assets/63248335/73d86f53-08ec-48c5-956b-129a0ca71c0a">

**Comment by jonathanbossenger** [comment]
Aha, so this is in the Post editor, but not the Site Editor? 

If so, do you think this tool makes sense in the context of developing a theme? 

Or better yet, if it does, how would you use it to test your theme template header levels?

**Comment by ironnysh** [comment]
It's also in the Site Editor (same place). I just randomly did a screenshot of the Post Editor :-)

**Comment by jonathanbossenger** [comment]
Aha! Got it, thank you!

**Comment by jonathanbossenger** [comment]
@ironnysh I'm recording this lesson, and I wanted to check with you, would you be ok if I split it into two lessons.

One focusing on the best practices, and then one focusing on the tools.

**Comment by ironnysh** [comment]
Sure, whatever makes sense.

This means we'll have 3 lessons for devs/designers.

**Comment by jonathanbossenger** [comment]
Markdown version of script for recording: [01-accessibility-best-practices.md](https://github.com/jonathanbossenger/wp-content/blob/main/theme-developer-learning-pathway/05-Accessiblity/01-accessibility-best-practices.md)

**Comment by ironnysh** [comment]
Hi @jonathanbossenger, I have a few comments on the script. What's the best way to do it? PR?

**Comment by jonathanbossenger** [comment]
Yup, a PR works.

**Comment by ironnysh** [comment]
Ready: https://github.com/jonathanbossenger/wp-content/pull/5

**Comment by jonathanbossenger** [comment]
Video for review: 

https://github.com/WordPress/Learn/assets/180629/de0b4704-47fb-424d-b4fa-16c6b5fef4ce

**Comment by jonathanbossenger** [comment]
https://wordpress.tv/2024/06/18/best-practices-for-developing-an-accessible-theme/

**Comment by jonathanbossenger** [comment]
https://youtu.be/zf1Y8s3a-Bk

**Comment by jonathanbossenger** [comment]
https://learn.wordpress.org/lesson/best-practices-for-developing-an-accessible-theme/

**Comment by jonathanbossenger** [comment]
//publish
